src/ILCompiler.Compiler/src/Compiler/DependencyAnalysis/ObjectNode.Sorting.cs

```diff
@@ -0,0 +1,144 @@
+            ArrayMapNode,
+            ReflectionFieldMapNode,
+            NativeLayoutInfoNode,
+            ExactMethodInstantiationsNode,
+            GenericTypesHashtableNode,
+            GenericMethodsHashtableNode,
+            GenericVirtualMethodTableNode,
+            InterfaceGenericVirtualMethodTableNode,
+            GenericMethodsTemplateMap,
+            GenericTypesTemplateMap,
+            BlockReflectionTypeMapNode,
+            StaticsInfoHashtableNode,
+            ReflectionVirtualInvokeMapNode,
+            ExternalReferencesTableNode,
+
+            //
+            // The ordering of this sequence of nodes is arbitrary.
+            //
+            TypeGenericDictionaryNode,
+            MethodGenericDictionaryNode,
+            ThreadStaticsIndexNode,
+            ThreadStaticsDirectoryNode,
+            TypeManagerIndirectionNode,
+            TypeThreadStaticIndexNode,
+            UnboxingStubNode,
+            UtcThreadStaticsNode,
+            MethodCodeNode,
+            InterfaceDispatchCellNode,
+            StandaloneGCStaticDescRegionNode,
+            RuntimeMethodHandleNode,
+            RuntimeFieldHandleNode,
+            ReadyToRunHeaderNode,
+            ReadyToRunGenericHelperNode,
+            PInvokeModuleFixupNode,
+            PInvokeMethodFixupNode,
+            NonGcStaticsNode,
+            NativeLayoutSignatureNode,
+            JumpStubNode,
+            NamedJumpStubNode,
+            ModulesSectionNode,
+            InterfaceDispatchMapNode,
+            InitialInterfaceDipatchStubNode,
+            IndirectionNode,
+            GenericCompositionNode,
+            GCStaticsPreInitDataNode,
+            GCStaticsNode,
+            GCStaticEETypeNode,
+            FatFunctionPointerNode,
+            EETypeOptionalFieldsNode,
+            EETypeNode,
+            ConstructedEETypeNode,
+            NecessaryCanonicalEETypeNode,
+            CanonicalEETypeNode,
+            BlobNode,
+            EmbeddedDataContainerNode,
+            ArrayOfEmbeddedPointersNode,
+            ReadyToRunHelperNode,
+            ArrayOfFrozenObjectsNode,
+        }
+
+        public class Comparer : IComparer<DependencyNodeCore<NodeFactory>>
+        {
+            private TypeSystemComparer _comparer;
+            private NameMangler _nameMangler;
+
+            public Comparer(TypeSystemComparer comparer, NameMangler nameMangler)
+            {
+                _comparer = comparer;
+                _nameMangler = nameMangler;
+            }
+            
+            public int Compare(DependencyNodeCore<NodeFactory> x1, DependencyNodeCore<NodeFactory> y1)
+            {
+                ObjectNode x = x1 as ObjectNode;
+                ObjectNode y = y1 as ObjectNode;
+
+                if (x == y)
+                {
+                    return 0;
+                }
+
+                // Sort non-object nodes after ObjectNodes  <TODO: Is this okay for determinism? Ideally we don't emit any non-ObjectNodes>
```

Comment:
The graph contains nodes we don't emit, ie `MethodMetadataNode` that derive from `DependencyNodeCore<NodeFactory>` directly. We're asked to compare them with ObjectNodes that we will emit so I needed some way deal with them.